### PR TITLE
lint: add new deferunlockcheck pass

### PR DESCRIPTION
**Pull request by Santamaura**
This commit adds a new pass which checks for
`...Unlock()` expressions without `defer` which
could result in a lock being held indefinitely.

Resolves #105366

Release note: None

### CODE REVIEW

**Review by knz — COMMENTED** [review]
could you also arrange for the check to be disabled if there's a special `// nolint:XXX` comment next to an Unlock call to say to disable the linter? 

See the other linters - we have `// nolint:errwrap`, `// nolint:errcmp` etc.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107577#-:-NaGit2B8P9oAwbRJRcH:b-yk4qd6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

**Review by Santamaura — COMMENTED** [review]
The check does account for this :smile: the exclusion rule is `// nolint:deferunlock`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107577#-:-NaHd4kGAvrf1IsVm7LC:b-94vao4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

**Review by rickystewart — APPROVED** [review]
Looks like CI is failing, have a look.

**Review by rickystewart — COMMENTED** [review]
Oh, also, add `//pkg/testutils/lint/passes/deferunlockcheck` along with the other lint checks in the list in top-level `BUILD.bazel` as part of `nogo`.

**Comment by rickystewart** [comment]
Looks like you will have to go into `build/bazelutil/nogo_config.json` and add a stanza for `deferunlockcheck` so that we only check first-party code. You can copy one of the existing blocks. Let me know if you have any questions.

**Review by Santamaura — COMMENTED** [review]
added
```
"deferunlockcheck": {

"only\_files": {

"cockroach/pkg/.\*$": "first-party code",

"cockroach/bazel-out/.\*/bin/pkg/.\*$": "first-party code"

}

},
```
to the `nogo_config.json` but build still seems to fail on external code..anything additional needed?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107577#-:-NaI-GkS4wx05uPcdKId:bypa8s1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

**Comment by rickystewart** [comment]
What's going on with the backslash?

```
"only\_files":
```

**Review by Santamaura — COMMENTED** [review]
Odd it wasn't a typo, it's inserted when copy-pasting


> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107577#-:-NaI1I5327yzV9uc25Bo:baxl2rm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

**Comment by rickystewart** [comment]
You spelled the name of your linter wrong. It's called `deferunlockcall`, not `deferunlockcheck`.

This diff works. Should you copy-paste it, be very careful you do not insert backslashes.

```
diff --git a/build/bazelutil/nogo_config.json b/build/bazelutil/nogo_config.json
index cddfbc90d5e..b5e9b80b09e 100644
--- a/build/bazelutil/nogo_config.json
+++ b/build/bazelutil/nogo_config.json
@@ -28,6 +28,12 @@
             "_test\\.go$": "tests"
         }
     },
+    "deferunlockcall": {
+        "only_files": {
+            "cockroach/pkg/.*$": "first-party code",
+            "cockroach/bazel-out/.*/bin/pkg/.*$": "first-party code"
+        }
+    },
     "errcheck": {
         "exclude_files": {
             "pkg/.*\\.eg\\.go$": "generated code",

```

**Comment by rickystewart** [comment]
It looks like you use `deferunlockcheck` in most of the places besides the actual name of the Analyzer. Is that intentional? You may find it easier just to update the Analyzer name to `deferunlockcheck` if that's what you *want* it to be called.

**Comment by rickystewart** [comment]
Oh, and the PR calls it `deferunlocktest`. :P It will probably be much less confusing for everyone if we select one name.

**Review by Santamaura — COMMENTED** [review]
Yeah that was the problem thanks! In this scenario should I `nolint` the current offenders to pass CI?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107577#-:-NaIYbJiELchun0zEsUl:boh3bai)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

**Comment by rickystewart** [comment]
> In this scenario should I nolint the current offenders to pass CI?

If you do, at least file bugs so we can keep track of the current offenders. Otherwise they'll just get lost.

**Comment by Santamaura** [comment]
There is a group effort internally right now to address unsafe locks so I will hold off on merging this until that work progresses since it will resolve the problems this linter is picking up anyways.

**Comment by Santamaura** [comment]
@rickystewart I have PRs that are addressing https://github.com/cockroachdb/cockroach/issues/107946 but there will still be many manual `Unlock`s which this linter will catch, is there a way to set the linter such that it would only capture new cases or is the only way to `nolint` all the remaining safe cases?

**Comment by Santamaura** [comment]
OK per discussion with @dhartunian at our retro, we don't want to `nolint` all manual `Unlocks()` since there are definitely some cases that should to be fixed but also figuring which ones those are and fixing them is gonna be considerable effort (and likely not gonna be done in the short term). Some cases are covered in #108064, #108171, and #108218 but there are still many more, Changing the linter to report after a certain number of manual unlocks will enforce future behavior to either `defer` or use the `nolint`. Gonna file some issues organized by team to address at a later time.

**Comment by Santamaura** [comment]
> Gonna file some issues organized by team to address at a later time.
Done. Will merge this eod

File: pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go
Comment (rickystewart) [inline pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go:62]: What is this supposed to do? If a given package has more than 1037 manual locks/unlocks it is an error? What is the significance of this number?

**Review by rickystewart — CHANGES_REQUESTED** [review]
This is not the way to solve this problem. As mentioned above, the way to solve it is:

1. tag the existing failures with `nolint`
2. File bugs for `nolint` comments
3. Over time, we'll whittle down the list of exclusions, or keep them tagged `nolint` in the minority of cases where that logic is "correct" (you don't have to do this all yourself, these tasks can be triaged to code owners)

**Comment by dhartunian** [comment]
@rickystewart perhaps I should explain my perspective here: I'm trying to strike a balance of immediate impact while also not introducing low-value toil. I understand that your perspective is in that it honors the linter's demands.

I'm skeptical that adding 1000s of `// nolint` comments and filing bugs will improve the situation in the near/medium term. There's little motivation to do this because there's not an identified problem db-wide of "unsafe unlocks" leading to customer issues. The problems are very localized to certain subsystems that either have low test coverage or don't see much production usage. 

Plus I'm a little nervous that adding lint-skip comments in lots of sensitive code will reduce readability. Today we have 136 instances of `// nolint` in our codebase which is why I'm nervous of increasing this number by an order of magnitude.

I do think that **preventing** new unsafe unlocks from entering the codebase is a strictly good thing and is low-impact because it will catch the problem at time of PR which will lead to engineers fixing the problem or adding a comment skipping the linter with an explanation.

My intention with the magic number that @Santamaura was mentioning was to count the existing number of violations and simply prevent our count from increasing, which will catch new problems without requiring that we address the existing giant backlog.

**Comment by knz** [comment]
> My intention with the magic number that @Santamaura was mentioning was to count the existing number of violations and simply prevent our count from increasing, which will catch new problems without requiring that we address the existing giant backlog.

The unintended risk here is that new violations get introduced while others get removed.

We could combine this idea of a count with a list of source files where the violations are allowed.

**Comment by rickystewart** [comment]
> My intention with the magic number that @Santamaura was mentioning was to count the existing number of violations and simply prevent our count from increasing, which will catch new problems without requiring that we address the existing giant backlog.

The proposed "solution" doesn't even accomplish this goal. I suspect the thinking is that the Analyzer will count every instance of the bad pattern in the entire codebase. Of course, it doesn't work like that: Analyzers work on a package-by-package basis. Raphael is right that even if it worked this way, this would mean that we can add new "violations" as we remove old ones, with no one being any the wiser. And of course no one will go and decrement the 1037 constant by exactly the correct number as they remove "unsafe unlocks".

> Plus I'm a little nervous that adding lint-skip comments in lots of sensitive code will reduce readability. Today we have 136 instances of // nolint in our codebase which is why I'm nervous of increasing this number by an order of magnitude.

You can put file-level `excludes` in `nogo_config.json` to avoid polluting the code if that's the concern.

> There's little motivation to do this because there's not an identified problem db-wide of "unsafe unlocks" leading to customer issues.

Okay, so what are we doing here? 😛 Is this linter solving a problem or not? If it's not solving a problem, let's just close the PR now. A linter that is basically guaranteed to not catch any errors is just wasting everyone's time (developer time, CI time), not to mention confusing for people trying to wrap their heads around the build system.

**Comment by rickystewart** [comment]
All of this is basically to say: if we're going to do this, let's do it right. If we're not going to do it right, why are we bothering?

**Comment by Santamaura** [comment]
From our weekly sync, I'm gonna first figure out how many occurrences of "problematic" lock statements there are (locks that are >5 lines away from the unlock) and we can evaluate further from there.

**Comment by Santamaura** [comment]
> From our weekly sync, I'm gonna first figure out how many occurrences of "problematic" lock statements there are (locks that are >5 lines away from the unlock) and we can evaluate further from there.

OK so I made an `awk` script that did this reasonably well, did some analysis with @knz yesterday and the conclusion is the following:
If you look at the histograms below, there are 124 "problematic" files with the majority of the "problematic" unlocks being between 5-10 lines away (I counted ~72). If we increase our criteria to >10 lines we only have 52 files to deal with and that should be an acceptable number to exclude in the `no_go` and create follow up issues for. So our plan is that we will apply the `nolint` to occurrences <= 10 lines which people should try to fix as they come across them in their daily activities and then the remainder will be excluded in the `no_go` and have issues filed for them. All this assumes the script was perfect which I doubt it was so there may be some error of margin in the actual number of files. Does this sound reasonable?

Screenshot 2023-08-14 at 4 58 09 PM
Screenshot 2023-08-14 at 5 18 47 PM

**Comment by knz** [comment]
sgtm

**Comment by Santamaura** [comment]
76 files to exclude is a decent amount higher than we thought but my open PRs should bring that number down some

**Comment by miretskiy** [comment]
> This commit adds a new pass which checks for `...Unlock()` expressions without `defer` which could result in a lock being held indefinitely.
> 
> Resolves #105366
> 
> Release note: None
Help me understand: why using the API, ie calling unlock is a bad thing? How is Unlock call without defer constitutes a bug? How would you use this API in a loop?

Seems to me that a way to detect misuse is to run tests with deadlock detection.

**Comment by knz** [comment]
> Help me understand: why using the API, ie calling unlock is a bad thing? How is Unlock call without defer constitutes a bug? How would you use this API in a loop?

There are two problems.

- The main problem is that when the Unlock call is "far" from the Lock, it's easy for other engineers who come to the code to maintain/extend it to forget to notice there is a Lock/Unlock pair; and then do things in-between that are not valid or are unsafe.
- The other problem is that if there is a panic in-between, the Unlock doesn't run. Since we catch and handle panics gracefully in many cases, a missing Unlock will cause a deadlock later.  This typically cannot be found using the deadlock detector during testing because we do not test all the situations that can lead to a panic.

**Comment by knz** [comment]
@Santamaura I invite you to add the motivation above to the commit message (and PR description), so that folk coming to this later via the git history understand why this came up.

**Comment by miretskiy** [comment]
> * The main problem is that when the Unlock call is "far" from the Lock, it's easy for other engineers who come to the code to maintain/extend it to forget to notice there is a Lock/Unlock pair; and then do things in-between that are not valid or are unsafe.

This sounds like a style problem to me, more than correctness problem.  
Sure, bugs do happen.  It's just calling Unlock directly is perfectly fine, and imo should be used more often (at least in my opinion) -- for example in simple setters such as the one below.
```
func (x *foo) setFooLocked(n int)  {
  x.Lock()
  x.foo = n
  x.Unlock()
}
```
It's not clear what defer buys you -- and, I think at least some time in the past, putting defer() in this function
would make it non-inlinable.

Then, you have totally legitimate cases where you use the lock in a loop - perhaps around sync.Cond; 
The problems of lock mis-use are legitimate, but I really don't believe that prescriptive solution like this
is a way to go since it prevents or at least makes rather annoying, legitimate uses of low level synchronization
primitives.

Personally, I'd be 👎 on this proposed lint.  I wonder what @dt and @stevendanna thinks.

**Comment by knz** [comment]
> This sounds like a style problem to me, more than correctness problem.

Agreed, hence the recommendation to use `defer` as a style choice.

> (regarding the example of a small accessor) It's not clear what defer buys you

In this particular example it doesn't buy much, but consider this other more interesting example:

```go
func (x *foo) setFooLocked(n int)  {
  x.Lock()
  *x.foo = n
  x.Unlock()
}
```

This will not behave correctly if `foo` happens to be `nil`. If there was a defer, the lock would be released properly (especially important if there's a panic catcher up the stack).

> and, I think at least some time in the past, putting defer() in this function would make it non-inlinable.

FUD - this was fixed in go 1.15.

> Then, you have totally legitimate cases where you use the lock in a loop - perhaps around sync.Cond;

There are a few things the style guide will offer for this:
- put the body of the loop into a separate function that uses defer
- use an inline function `func() { ... }()` inside the body
- use a `// nolint:deferunlock` annotation

> I really don't believe that prescriptive solution like this is a way to go since it prevents or at least makes rather annoying, legitimate uses of low level synchronization primitives.

Yes, a linter is intended to make certain things annoying. That is exactly the point. There are, indeed, _some_ legitimate uses of non-deferred unlock, which is why there will be an opt-out. The experience has shown however that *in the common case* we do not wish to use non-deferred unlock, because they introduce bugs.

**Comment by miretskiy** [comment]
> FUD - this was fixed in go 1.15.

How is it fud if it was the case before? I'm just saying there were reasons not to do it.

> Yes, a linter is intended to make certain things annoying. That is exactly the point. 

I disagree --  a linter that fires off all the time will simply teach to put `// nolint` comment all the time, defeating the purpose.  The use of locks in the loop are *very* common (any sync.Cond, afaik); Defer in a loop -- now that's a bug and this lint will encourage me to put defer in a loop.


> There are a few things the style guide will offer for this:
> 
> * put the body of the loop into a separate function that uses defer
> * use an inline function `func() { ... }()` inside the body
> * use a `// nolint:deferunlock` annotation

Sure -- but that's exactly my point -- we have style guide for this problem, we don't need lint.
The fact that we don't "enforce" style guid during reviews, that's a different issue, I think.



> In this particular example it doesn't buy much, but consider this other more interesting example:
> 
> ```go
> func (x *foo) setFooLocked(n int)  {
>   x.Lock()
>   *x.bar = n
>   x.Unlock()
> }
> ```
> 
> This will not behave correctly if `bar` happens to be `nil`. If there was a defer, the lock would be released properly (especially important if there's a panic catcher up the stack).

Why are we conflating nil ptr issues with lock/unlock issue?  Whether or not you panic and cause process to die, or you panic and recover (and cause deadlock later) -- you still have a bug, no?  Why are we trying to pretend that NPE has anything to do w/  locking?

**Review by stevendanna — COMMENTED** [review]
Thanks for working on tooling to improve things.

In the second commit we say:

> The commit also applies the nolint rule to any remaining manual unlock cases that are <=10 lines away from the lock.

Would it be possible to build this into the linter? My guess is that a many of the cases people may object to changing would be OK if we ignored non-defered-Unlock()'s that were <= 5 lines away from the Lock().

I reviewed some of the call sites for the code I know and I think such an exception would have handled many of the cases where I think the code as-is would be OK. And, an advantage is that we could avoid the nolint directive in a lot of cases, which means that if the logic inside the locked region grows, we'll get a fresh lint failure when it does.

We'd have to do something where we track the Lock() calls and record their positions. I think this is possible but haven't worked with the linter tooling in a little while to know how much work it is off hand.

I know at some point we were worried about the overhead of `defer` in hot code paths. Is that still the case? I threw some small examples into godbolt and you get a bit more assembly for defer, but perhaps that doesn't turn into a difference we care about. 

> Help me understand: why using the API, ie calling unlock is a bad thing? How is Unlock call without defer constitutes a bug?

@miretskiy One argument for this isn't that Unlock without defer is a _bug_; but rather, that it can lead to bugs down the road.

That is, a new simple thing that gets added and it looks like this:

```golang
mu.Lock()
whatever = foo
mu.Unlock()
```

and it is obviously correct and everyone is happy. And then maybe tomorrow it becomes:

```golang
mu.Lock()
// A few more things....
//
// And maybe also a big comment because we love comments.
whatever = foo
//
// A few more things, each added for a reason.
// 
mu.Unlock()
```

Things are still correct, but we are a little less happy because now the Unlock is far away from Lock().

And then, suddenly, we need to do this:

```golang
mu.Lock()
// A few more things....
//
// And maybe also a big comment because we love comments.
var err error
whatever, err = makeWhatever(foo)
if err != nil {
    return err
}
//
// A few more things, each added for a reason.
// 
mu.Unlock()
```

And maybe the reviewer didn't expand the diff's context enough during review to see that this was all happening under the lock. Now, things are not correct and we are unhappy. 

There is of course the opposite problem where you use `defer mu.Unlock()` and then that function grows and you add some long-running operation that should not be happening under a lock.

**Comment by miretskiy** [comment]
> Thanks for working on tooling to improve things.
> 
> In the second commit we say:
> 
> > The commit also applies the nolint rule to any remaining manual unlock cases that are <=10 lines away from the lock.
> 
> Would it be possible to build this into the linter? My guess is that a many of the cases people may object to changing would be OK if we ignored non-defered-Unlock()'s that were <= 5 lines away from the Lock().
> 
> I reviewed some of the call sites for the code I know and I think such an exception would have handled many of the cases where I think the code as-is would be OK. And, an advantage is that we could avoid the nolint directive in a lot of cases, which means that if the logic inside the locked region grows, we'll get a fresh lint failure when it does.
> 
> We'd have to do something where we track the Lock() calls and record their positions. I think this is possible but haven't worked with the linter tooling in a little while to know how much work it is off hand.
> 
> I know at some point we were worried about the overhead of `defer` in hot code paths. Is that still the case? I threw some small examples into godbolt and you get a bit more assembly for defer, but perhaps that doesn't turn into a difference we care about.
> 
> > Help me understand: why using the API, ie calling unlock is a bad thing? How is Unlock call without defer constitutes a bug?
> 
> @miretskiy One argument for this isn't that Unlock without defer is a _bug_; but rather, that it can lead to bugs down the road.
> 
> That is, a new simple thing that gets added and it looks like this:
> 
> ```go
> mu.Lock()
> whatever = foo
> mu.Unlock()
> ```
> 
> and it is obviously correct and everyone is happy. And then maybe tomorrow it becomes:
> 
> ```go
> mu.Lock()
> // A few more things....
> //
> // And maybe also a big comment because we love comments.
> whatever = foo
> //
> // A few more things, each added for a reason.
> // 
> mu.Unlock()
> ```
> 
> Things are still correct, but we are a little less happy because now the Unlock is far away from Lock().
> 
> And then, suddenly, we need to do this:
> 
> ```go
> mu.Lock()
> // A few more things....
> //
> // And maybe also a big comment because we love comments.
> var err error
> whatever, err = makeWhatever(foo)
> if err != nil {
>     return err
> }
> //
> // A few more things, each added for a reason.
> // 
> mu.Unlock()
> ```
> 
> And maybe the reviewer didn't expand the diff's context enough during review to see that this was all happening under the lock. Now, things are not correct and we are unhappy.
> 
> There is of course the opposite problem where you use `defer mu.Unlock()` and then that function grows and you add some long-running operation that should not be happening under a lock.

@stevendanna  -- that also sounds like we need to do better job during code reviews to encourage good style
patterns, including refactoring large functions.

**Comment by stevendanna** [comment]
Apologies that my comment seems to ignore the previous commentary. I wrote it while you all were talking and didn't refresh the page before submitting.

**Comment by miretskiy** [comment]
Serious question:  can we run some process (like blathers) once PR is made to add comments to the PR
to say that "it's good *style* to put `defer x.Unlock()` immediately after x.Lock()"?

Make automatic PR comments -- let the reviewer and the author sort things out.  No need for lint that
prevents legitimate uses.

**Comment by JeffSwenson** [comment]
I'm strongly in favor of a linter that requires defering unlock. Non-deferred unlocks are a trivial to avoid source of unwind unsafe code. On the CC team we had a production impacting deadlock that was caused by a raw unlock in the go stdlib, so I don't think this is a case where sufficiently smart programmers can avoid making the mistake. It can be a difficult thing to catch in code review because the changes that make code unsafe often occur over several disconnected PRs.

It is true that Defer prevents inlinining. But the performance impact is usually small because it doesn't defeat unwind analysis. I also suspect the register based calling conventions make small functions nearly free to call. 
https://cockroachlabs.slack.com/archives/CLQ753JGH/p1691678723353409?thread_ts=1691629384.001819&cid=CLQ753JGH
https://cockroachlabs.slack.com/archives/C0HM2DZ34/p1687453282632779?thread_ts=1687450860.116279&cid=C0HM2DZ34

I also think @miretskiy's concern around widespread use of nolint is valid. If we do allow undeferred unlocks, we could augment the linter with automatic exceptions that are relatively strict. E.g. no more than 5 lines of code between the lock and the unlock. No function or method calls in the critical section. That covers the examples I have seen of reasonable non-deferred unlocks.

**Comment by miretskiy** [comment]
> I'm strongly in favor of a linter that requires defering unlock.

Would a stronger emphasis on reviews that insists on defer unlock be sufficient?

**Comment by knz** [comment]
> Would a stronger emphasis on reviews that insists on defer unlock be sufficient?

No. This line of thinking fails to acknowledge why linters exist in the first place. If we bought into the idea that "reviews are sufficient" we would not need linters _at all_ - reviewers would just pick everything up.

The reality is that there are many things that "look small" during reviews (and thus escape scrutiny) but have *large impact* in practice. For example, mistakenly importing `"errors"` instead of `"github.com/cockroachdb/errors"` - it *appears* to work but creates hard-to-troubleshoot real bugs.

Uses of `Unlock` without `defer` are in this category, it looks small but has large impact down the line.

I do agree with Steven however, that we probably want a smart(er) linter which automatically avoids the warning if the Lock is "close enough" to the Unlock call.

**Comment by miretskiy** [comment]
> No. This line of thinking fails to acknowledge why linters exist in the first place. If we bought into the idea that "reviews are sufficient" we would not need linters _at all_ - reviewers would just pick everything up.
> 

Well, that's taking a bit of an extreme approach.  Linters are useful.  The fact is: locks are low level intrinsic part of the language; they can be used and they can be mis-used.  How is misusing channel (i.e close it twice) any different from mis-using a lock?

If I'm new to this codebase, and I grep for uses of Unlock, I will 
find that 291 files use unlock with "nolint" -- I'll conclude that that's they way it ought to work.
This lint is not targeted, has more false positives than potential problem it tries to solve, and overall just makes
things worse -- readability wise.

> The reality is that there are many things that "look small" during reviews (and thus escape scrutiny) but have _large impact_ in practice. For example, mistakenly importing `"errors"` instead of `"github.com/cockroachdb/errors"` - it _appears_ to work but creates hard-to-troubleshoot real bugs.
> 
> Uses of `Unlock` without `defer` are in this category, it looks small but has large impact down the line.
> 
> I do agree with Steven however, that we probably want a smart(er) linter which automatically avoids the warning if the Lock is "close enough" to the Unlock call.

Sure, we can do that; but then of course, I don't need many lines to panic -- which again, begs the question: why bother with this lint?

A blathers or similar system that puts automatic comments on the PR to bring attention to the fact that lock does not follow with "defer Unlock" seems strictly better.

**Comment by knz** [comment]
> This lint is not targeted, has more false positives than potential problem it tries to solve, and overall just makes
things worse -- readability wise.

I think this point is well taken and a sign we should make this linter slightly more clever.

>  but then of course, I don't need many lines to panic

That is another thing the linter can do: only flag lock/unlock pair that have either:
- non-linear control flow in-between (i.e. flag if there's a condition, function call or a loop in the middle).
- access to nullable fields in data structures.

**Comment by rafiss** [comment]
> Make automatic PR comments -- let the reviewer and the author sort things out. No need for lint that prevents legitimate uses.

IMO a linter is exactly the right place to "start a conversation" between the reviewer and author about what needs to be sorted out. I don't think automatic PR comments are as effective at that. 

Seems like the major concern here would be resolved if we made sure there were fewer `nolint` exceptions before merging this.

**Comment by renatolabs** [comment]
I'm also in favour of some form of linting for this pattern, thanks for working on it.

Others already covered the reasons why I think it's important to have tooling to stop these kind of errors instead of relying on human reviews so I won't repeat that; at the end of the day, people are fallible and we _will_ miss issues in the code when it's sufficiently complex.

I also agree with the general sentiment that we could make the linter a little smarter. Things like:

> non-linear control flow in-between (i.e. flag if there's a condition, function call or a loop in the middle).

> no more than 5 lines of code between the lock and the unlock. No function or method calls in the critical section

make sense to me and would help us focus on the more likely to be problematic "unsafe" uses of `Unlock`.

> We'd have to do something where we track the Lock() calls and record their positions. I think this is possible but haven't worked with the linter tooling in a little while to know how much work it is off hand.

This shouldn't be too much work. We can keep track of which object is being locked and unlocked in a way that is similar to [1]. Line numbers can also be computed by calling `Position` on the underlying `*token.FileSet` [2]. Feel free to ping me if working on something like this, @Santamaura, I've used these Analyzer APIs before and would be happy to help if I can. 

[1] https://github.com/cockroachdb/cockroach/blob/75c6056aad3574afef45888533c3df9133ddf65d/pkg/testutils/lint/passes/loopvarcapture/loopvarcapture.go#L109
[2] https://pkg.go.dev/go/token#FileSet.Position

File: pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go
Comment (renatolabs) [inline pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go:53]: We probably want to make sure we're only reporting issues with `Unlock` calls called on instances of `sync{,util}.{RW}Mutex`. Currently if a custom struct implements an `Unlock` function, the linter will complain about it and we probably don't want that, as the semantics are unknown to the linter.

Some similar code:

https://github.com/cockroachdb/cockroach/blob/75c6056aad3574afef45888533c3df9133ddf65d/pkg/testutils/lint/passes/loopvarcapture/loopvarcapture.go#L404-L407

**Comment by renatolabs** [comment]
The linter will also currently complain about code like below, and it probably shouldn't:

```go
mu.Lock()
defer func() {
	cancel()
    mu.Unlock()
}()
```

**Comment by Santamaura** [comment]
Made the "smarter" linter but still doing some testing. Need to update the `nogo_config` once done with testing.

**Comment by Santamaura** [comment]
OK this should be RFAL

**Comment by Santamaura** [comment]
Should also mention my team wants me to timebox any more time spent on this linter since it's not really obs inf domain so if this iteration ain't it then I'll probably need to hand it off to dev inf.

**Review by stevendanna — COMMENTED** [review]
Thanks for taking time to address the concerns around this linter.

For me, the inclusion of the "close enough" exception makes this something worth trying.

I'm not convinced we need to report an error on for/range statements. but can live with it if y'all feel it is important.

I briefly read over the linter code, but it should probably have some review from someone more familiar with the relevant APIs.

**Comment by Santamaura** [comment]
Not too sure who to get a review from maybe @renatolabs since your familiar with the analyzer apis?

File: pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go
Comment (rickystewart) [inline pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go:298]: Seems like you can immediately `if hasNoLintComment { return false }`, right? We will always return false in this case with the way this function is currently written.

File: pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go
Comment (rickystewart) [inline pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go:40]: Nit: `noLintName` and `maxLineDistance` should be `const`, not `var`.

File: pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go
Comment (rickystewart) [inline pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go:286]: The error message here (and on line 311) is not great. Note that this is the message that will be displayed to devs when they violate the lint rule. In particular it's unclear what someone is supposed to do to solve the problem. e.g. "Oh, I can't have more than 5 lines between the lock and unlock, so I need to start combining multiple lines of code into one line with semicolons" or something like that.

The error message should succinctly describe what the actual problem is and suggest a fix like "move the `Unlock()` call to a `defer` statement immediately after the `Lock()` call".

The error message on line 311 has an additional problem because what is "non-linear control flow"? Why is it a problem if I have non-linear control flow, and how do I fix it?

**Review by rickystewart — CHANGES_REQUESTED** [review]

**Review by Santamaura — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107577)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling, @renatolabs, and @rickystewart)

___
*[`pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go` line 40 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107577#gh-1312039468:-Nd_kiDOBoMh_Jd5kxI2:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/70f1b241c1e6880d81daa38aaf8abcc0418a1b8e/pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go#L40)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

Nit: `noLintName` and `maxLineDistance` should be `const`, not `var`.
</blockquote></details>

Done
___
*[`pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go` line 286 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107577#gh-1312047308:-Nd_kyZb9ErcKggVOMw5:b-5lts0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/70f1b241c1e6880d81daa38aaf8abcc0418a1b8e/pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go#L286)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

The error message here (and on line 311) is not great. Note that this is the message that will be displayed to devs when they violate the lint rule. In particular it's unclear what someone is supposed to do to solve the problem. e.g. "Oh, I can't have more than 5 lines between the lock and unlock, so I need to start combining multiple lines of code into one line with semicolons" or something like that.

The error message should succinctly describe what the actual problem is and suggest a fix like "move the `Unlock()` call to a `defer` statement immediately after the `Lock()` call". If you produce something the developer can directly copy-paste (like `defer x.y.whatever.Unlock()` -- I think you actually already have the pieces you need to make this work), even better.

The error message on line 311 has an additional problem because what is "non-linear control flow"? Why is it a problem if I have non-linear control flow, and how do I fix it? 
</blockquote></details>

Thanks for the feedback, I made the error messages a bit simpler and added what the dev can do to fix it.
___
*[`pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go` line 298 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/107577#gh-1312034154:-Nd_kr2562IR5oP2ReO4:b-a36jdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/70f1b241c1e6880d81daa38aaf8abcc0418a1b8e/pkg/testutils/lint/passes/deferunlockcheck/deferunlockcheck.go#L298)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

Seems like you can immediately `if hasNoLintComment { return false }`, right? We will always return false in this case with the way this function is currently written.
</blockquote></details>

Agree, done.

**Review by rickystewart — APPROVED** [review]

**Comment by Santamaura** [comment]
TFTRs!

bors r+

**Pull request merged**
